src/os/kstore/KStore.cc

```diff
@@ -1898,10 +1898,13 @@ int KStore::_open_super_meta()
 
 void KStore::_assign_nid(TransContext *txc, OnodeRef o)
 {
-  if (o->onode.nid)
+  if (o->onode.nid) {
+    assert(o->exists);
     return;
+  }
   std::lock_guard<std::mutex> l(nid_lock);
   o->onode.nid = ++nid_last;
+  o->exists = true;
```

Comment:
I'd suggest to drop the necessary kstore change here, because:
(1) we are moving the assignment within the scope of nid_lock, which (slightly) increases the overhead here.
(2) It's almost dead, hence does not deserve the effort.